src/rpcrequest-payload.ts

```diff
@@ -64,26 +56,32 @@ class RpcRequestPayload {
     buffer.writeUInt16LE(optionFlags);
 
     const parameters = this.request.parameters;
-    const writeNext = (i: number) => {
-      if (i >= parameters.length) {
-        cb(buffer.data);
-        return;
-      }
+    const buffers: Buffer[] = [];
+
+    parameters.forEach((param: Parameter) => {
+      const gen = this.generateParameterData(param, this.options);
+      const next = () => {
+        const curr = gen.next();
+        if (curr.done) {
+          return;
+        }
+        buffers.push(curr.value);
+        next();
+      };
+      next();
+    });
+
+    buffer.writeBuffer(Buffer.concat(buffers));
 
-      this._writeParameterData(parameters[i], buffer, () => {
-        setImmediate(() => {
-          writeNext(i + 1);
-        });
-      });
-    };
-    writeNext(0);
+    yield buffer.data;
```

Comment:
I think by using `parameters.forEach`, you ran into a sort-of "limitation" of generator functions - that is, you can't `yield` from inside a closue.

I see you worked around this by consuming the generator and collecting all buffers in the `buffers` array first. This works, but there's a disadvantage here - we're back to having to collect all the request data we need to send off into a large buffer first, and we'll be back at having to pay with higher latency as well as higher memory usage.

A better solution would be to use a plain old `for` loop here, like this:

```js
// Yield data written to `buffer` so far
yield buffer.data;

// Yield parameter data
for(let i = 0; i < parameters.length; i++) {
  yield * this.generateParameterData(param, this.options);
}
```